File: 138-copy-list-with-random-pointer/138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        
        if (head == null) return null;
        
        Map<Node, Node> map = new HashMap<>();

        Node head2 = new Node(-1);
        Node prev = head2;
        Node cur = head;
        
        while(cur != null) {
            if (!map.containsKey(cur)) {
                Node temp = new Node(cur.val);
                map.put(cur, temp);
            }
            
            Node temp = map.get(cur);
            prev.next = temp;
            
            if (!map.containsKey(cur.random)) {
                Node tempRand = null;
                if (cur.random != null) tempRand = new Node(cur.random.val);
                map.put(cur.random, tempRand);
            }
            
            temp.random = map.get(cur.random);
            cur = cur.next;
            prev = prev.next;
        }
        
        Node travel = head2;
        head2 = head2.next;
        
        return head2;
    }
}

File: 138-copy-list-with-random-pointer/138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(head==NULL) return head;
        Node* cur=head;
        Node* head2= new Node(cur->val);
        cur=cur->next;
        Node* cur2=head2;
        while(cur){
            Node* temp= new Node(cur->val);
            cur=cur->next;
            cur2->next=temp;
            cur2=temp;
        }
        cur2->next=NULL;
        // Till now the linked list with next pointer is copied successfully,now just need to copy the random pointer brother.
        Node* trv= head;
        Node* trv2= head2;
        while(trv){
            Node* ahead=trv->next;
            trv->next=trv2;
            Node* achead=trv2->next;
            trv2->next=ahead;
            trv=ahead;
            trv2=achead;
            
        }
        // The two lists have been merged successfully;
        // ie head= 7->7->13->13->11->11->10->10->1->1->NULL
        Node* pass=head;
        while(pass){
            pass->next->random=pass->random ? pass->random->next: pass->random ;
            pass=pass->next->next;
        }
        // the list have been copied successfully now we will separate the two lists.
        Node* sep1=head;
        Node* sep2=head2;
        while(sep1->next->next){
            sep1->next=sep1->next->next;
            sep2->next=sep2->next->next;
            sep1=sep1->next;
            sep2=sep2->next;
        }
        sep1->next=NULL;
        return head2;
    }
};